model/riscv_insts_vext_vset.sail

```diff
@@ -0,0 +1,202 @@
+/*  2. Redistributions in binary form must reproduce the above copyright           */
+/*     notice, this list of conditions and the following disclaimer in             */
+/*     the documentation and/or other materials provided with the                  */
+/*     distribution.                                                               */
+/*                                                                                 */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''             */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED              */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A                */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR            */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,                   */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT               */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF               */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND            */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,             */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT             */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF             */
+/*  SUCH DAMAGE.                                                                   */
+/*=================================================================================*/
+
+/* ******************************************************************************* */
+/* This file implements part of the vector extension.                              */
+/* Chapter 6: Configuration-Setting Instructions                                   */
+/* ******************************************************************************* */
+
+mapping sew_flag : string <-> bits(3) = {
+  "e8"    <-> 0b000,
+  "e16"   <-> 0b001,
+  "e32"   <-> 0b010,
+  "e64"   <-> 0b011,
+  "e128"  <-> 0b100,
+  "e256"  <-> 0b101,
+  "e512"  <-> 0b110,
+  "e1024" <-> 0b111
+} 
+
+mapping maybe_lmul_flag : string <-> bits(3) = {
+  ""              <-> 0b000, /* m1 by default */
+  sep() ^ "mf8"   <-> 0b101,
+  sep() ^ "mf4"   <-> 0b110,
+  sep() ^ "mf2"   <-> 0b111,
+  sep() ^ "m1"    <-> 0b000,
+  sep() ^ "m2"    <-> 0b001,
+  sep() ^ "m4"    <-> 0b010,
+  sep() ^ "m8"    <-> 0b011
+}
+
+mapping maybe_ta_flag : string <-> bits(1) = {
+  ""           <-> 0b0, /* tu by default */
+  sep() ^ "ta" <-> 0b1,
+  sep() ^ "tu" <-> 0b0
+}
+
+mapping maybe_ma_flag : string <-> bits(1) = {
+  ""           <-> 0b0, /* mu by default */
+  sep() ^ "ma" <-> 0b1,
+  sep() ^ "mu" <-> 0b0
+}
+
+/* ****************************** vsetvli & vsetvl ******************************* */
+union clause ast = VSET_TYPE : (vsetop, bits(1), bits(1), bits(3), bits(3), regidx, regidx)
```

Comment:
Note that for ELEN=32 RISC-V architecture we have this note in the specification that seems to say sew is 2 bits and vill is encoded in sew. I hadn't noticed before because I have always worked with ELEN=64 architectures. I find it quite messy. Perhaps for this merge we can just add a check that V extension requires ELEN=64 and leave as a TODO if in case anyone requires ELEN=32 with v-ext to be modelled.

> A small implementation supporting ELEN=32 requires only seven bits of state in vtype: two bits for ma and ta, two bits for vsew[1:0] and three bits for vlmul[2:0]. The illegal value represented by vill can be internally encoded using the illegal 64-bit combination in vsew[1:0] without requiring an additional storage bit to hold vill.

Reply:
Got it.

Reply:
I have added a note in riscv_vlen.sail for this issue.